### PR TITLE
CloudNative.CloudEvents bridge library

### LINKED ISSUE #15588: Implement CloudEvent bridge library to integrate with CloudNative SDK
https://gist.github.com/JoshLove-msft/b2c1404c9bafaa60c0d5d4d63e26d5e9

### PR DESCRIPTION
**Pull request by JoshLove-msft**
Fixes https://github.com/Azure/azure-sdk-for-net/issues/15588

### CODE REVIEW

File: sdk/eventgrid/Microsoft.Azure.CloudNative.CloudEvents.EventGrid/README.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+## Contributing
+
+This project welcomes contributions and suggestions. Most contributions require you to agree to a Contributor License Agreement (CLA) declaring that you have the right to, and actually do, grant us the rights to use your contribution. For details, visit <https://cla.microsoft.com>.
+
+When you submit a pull request, a CLA-bot will automatically determine whether you need to provide a CLA and decorate the PR appropriately (e.g., label, comment). Simply follow the instructions provided by the bot. You will only need to do this once across all repos using our CLA.
+
+This project has adopted the [Microsoft Open Source Code of Conduct][code_of_conduct]. For more information see the [Code of Conduct FAQ][code_of_conduct_faq] or contact opencode@microsoft.com with any additional questions or comments.
+
+![Impressions](https://azure-sdk-impressions.azurewebsites.net/api/impressions/azure-sdk-for-net%2Fsdk%2Fcore%2FMicrosoft.Azure.Core.NewtonsoftJson%2FREADME.png)
```
Comment (ellismg) [inline sdk/eventgrid/Microsoft.Azure.CloudNative.CloudEvents.EventGrid/README.md:45]: NIT: Update path

File: sdk/eventgrid/Microsoft.Azure.CloudNative.CloudEvents.EventGrid/README.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+[code_of_conduct]: https://opensource.microsoft.com/codeofconduct
+[code_of_conduct_faq]: https://opensource.microsoft.com/codeofconduct/faq/
+[microsoft_spatial_package]: https://www.nuget.org/packages/Microsoft.Spatial/
```
Comment (ellismg) [inline sdk/eventgrid/Microsoft.Azure.CloudNative.CloudEvents.EventGrid/README.md:49]: NIT: Remove

File: sdk/eventgrid/Microsoft.Azure.CloudNative.CloudEvents.EventGrid/tests/CloudEventTests.cs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        private static List<CloudEvent> DeserializeRequest(Request request)
+        {
+            var content = request.Content;
+            var stream = new MemoryStream();
+            content.WriteTo(stream, CancellationToken.None);
+            stream.Position = 0;
+            JsonDocument requestDocument = JsonDocument.Parse(stream);
+            var cloudEvents = new List<CloudEvent>();
+            // Parse JsonElement into separate events, deserialize event envelope properties
+            if (requestDocument.RootElement.ValueKind == JsonValueKind.Object)
```
Comment (ellismg) [inline sdk/eventgrid/Microsoft.Azure.CloudNative.CloudEvents.EventGrid/tests/CloudEventTests.cs:121]: Can you help me understand the case where we get a RootElement.ValueKind of object here?  I thought that the top level object will always be an array (even if a single event is provided)?

**Review by ellismg — APPROVED** [review]
LGTM.  Love the documentation improvements as well :-)

File: sdk/eventgrid/Azure.Messaging.EventGrid/api/Azure.Messaging.EventGrid.netstandard2.0.cs [hunk 4]
```diff
@@ -49,6 +49,10 @@ public partial class EventGridPublisherClient
         public EventGridPublisherClient(System.Uri endpoint, Azure.Messaging.EventGrid.EventGridSharedAccessSignatureCredential credential) { }
         public EventGridPublisherClient(System.Uri endpoint, Azure.Messaging.EventGrid.EventGridSharedAccessSignatureCredential credential, Azure.Messaging.EventGrid.EventGridPublisherClientOptions options) { }
         public static string BuildSharedAccessSignature(System.Uri endpoint, System.DateTimeOffset expirationUtc, Azure.AzureKeyCredential key, Azure.Messaging.EventGrid.EventGridPublisherClientOptions.ServiceVersion apiVersion = Azure.Messaging.EventGrid.EventGridPublisherClientOptions.ServiceVersion.V2018_01_01) { throw null; }
+        [System.ComponentModel.EditorBrowsableAttribute(System.ComponentModel.EditorBrowsableState.Never)]
+        public virtual Azure.Response SendEncodedCloudEvents(System.ReadOnlyMemory<byte> cloudEvents, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
```
Comment (pakrym) [inline sdk/eventgrid/Azure.Messaging.EventGrid/api/Azure.Messaging.EventGrid.netstandard2.0.cs:53]: How do I receive an event sent this way?

File: sdk/eventgrid/Microsoft.Azure.CloudNative.CloudEvents.EventGrid/tests/Microsoft.Azure.CloudNative.CloudEvents.EventGrid.Tests.csproj [hunk 5]
```diff
@@ -0,0 +1,23 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFrameworks>$(RequiredTargetFrameworks)</TargetFrameworks>
+
+    <!-- We don't care about XML doc comments on test methods -->
+    <NoWarn>$(NoWarn);CS1591</NoWarn>
+  </PropertyGroup>
+
+  <!-- Reference the Client Library -->
```
Comment (pakrym) [inline sdk/eventgrid/Microsoft.Azure.CloudNative.CloudEvents.EventGrid/tests/Microsoft.Azure.CloudNative.CloudEvents.EventGrid.Tests.csproj:9]: This comment is alone.

File: sdk/eventgrid/Azure.Messaging.EventGrid/api/Azure.Messaging.EventGrid.netstandard2.0.cs (hunk 4 above)
Comment (JoshLove-msft) [reply sdk/eventgrid/Azure.Messaging.EventGrid/api/Azure.Messaging.EventGrid.netstandard2.0.cs:53]: Receiving would be the same as for any other event that gets published to a topic.

File: sdk/eventgrid/Microsoft.Azure.CloudNative.CloudEvents.EventGrid/src/EventGridPublisherClientExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+            string traceState = null;
+            Activity currentActivity = Activity.Current;
+            if (currentActivity != null && currentActivity.IsW3CFormat())
+            {
+                activityId = currentActivity.Id;
+                currentActivity.TryGetTraceState(out traceState);
+            }
+
+            using var stream = new MemoryStream();
+            var writer = new Utf8JsonWriter(stream);
```
Comment (pakrym) [inline sdk/eventgrid/Microsoft.Azure.CloudNative.CloudEvents.EventGrid/src/EventGridPublisherClientExtensions.cs:69]:
```suggestion
            using var writer = new Utf8JsonWriter(stream);
```

File: sdk/eventgrid/Microsoft.Azure.CloudNative.CloudEvents.EventGrid/tests/CloudEventTests.cs [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+namespace Microsoft.Azure.CloudNative.CloudEvents.EventGrid.Tests
+{
+    public class CloudEventTests
+    {
+
+        private const string TraceParentHeaderName = "traceparent";
+        private const string TraceStateHeaderName = "tracestate";
+
+        private static readonly JsonEventFormatter s_eventFormatter = new JsonEventFormatter();
+
```
Comment (pakrym) [inline sdk/eventgrid/Microsoft.Azure.CloudNative.CloudEvents.EventGrid/tests/CloudEventTests.cs:29]: nit: ws
```suggestion
```

File: sdk/eventgrid/Microsoft.Azure.CloudNative.CloudEvents.EventGrid/tests/CloudEventTests.cs [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+            await client.SendCloudEventsAsync(eventsList);
+
+            activity.Stop();
+            List<CloudEvent> cloudEvents = DeserializeRequest(mockTransport.SingleRequest);
+            IEnumerator<CloudEvent> cloudEnum = eventsList.GetEnumerator();
+            foreach (CloudEvent cloudEvent in cloudEvents)
+            {
+                cloudEnum.MoveNext();
+                IDictionary<string, object> cloudEventAttr = cloudEnum.Current.GetAttributes();
+                if (cloudEventAttr.ContainsKey(TraceParentHeaderName) &&
```
Comment (pakrym) [inline sdk/eventgrid/Microsoft.Azure.CloudNative.CloudEvents.EventGrid/tests/CloudEventTests.cs:80]: use `inclTraceparent` and `inclTracestate` for conditions here.

**Review by pakrym — APPROVED** [review]

File: sdk/eventgrid/Microsoft.Azure.CloudNative.CloudEvents.EventGrid/tests/CloudEventTests.cs (hunk 3 above)
Comment (JoshLove-msft) [reply sdk/eventgrid/Microsoft.Azure.CloudNative.CloudEvents.EventGrid/tests/CloudEventTests.cs:121]: This was carried forward from the Track 1 library deserialization handling. It is definitely not needed here as the input is the array that was constructed when publishing. Will remove.

File: sdk/eventgrid/Microsoft.Azure.CloudNative.CloudEvents.EventGrid/tests/CloudEventTests.cs (hunk 8 above)
Comment (JoshLove-msft) [reply sdk/eventgrid/Microsoft.Azure.CloudNative.CloudEvents.EventGrid/tests/CloudEventTests.cs:80]: I have that i%2 check so it is not enough to check those flags.

**Pull request merged**